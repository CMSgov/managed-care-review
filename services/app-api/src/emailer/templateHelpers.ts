import * as Eta from 'eta'
import * as path from 'path'

import {
    LockedHealthPlanFormDataType,
    UnlockedHealthPlanFormDataType,
    SubmissionType,
} from '../../../app-web/src/common-code/healthPlanFormDataType'
import { EmailConfiguration, StateAnalystsEmails } from '.'
import { ProgramType } from '../domain-models'
import { logError } from '../logger'
import { pruneDuplicateEmails } from './formatters'

// ETA SETUP
Eta.configure({
    cache: true, // Make Eta cache templates
    views: [
        path.join(__dirname, 'etaTemplates'),
        path.join(__dirname, '../../'), // this is intentional - we need to have this path here to be able to find the files from the graphql lambda when in serverless deployed app
    ],
})

const renderTemplate = async <T extends object>(
    templateName: string,
    data: T
) => {
    if (!/^[a-zA-Z0-9]+$/.test(templateName)) {
        console.error(
            'CODING ERROR: templateName parameter should not include any punctuation, can only be alphanumeric characters'
        )
        return new Error(`${templateName} is not a valid template file name`)
    }
    const templatePath = `./${templateName}`

    try {
        const templateOrVoid = await Eta.renderFile(templatePath, data)

        if (typeof templateOrVoid !== 'string') {
            return new Error(
                `Could not render file ${templatePath}, no template string returned`
            )
        }
        const templateHTML = templateOrVoid as string // we know we have a string we can coerce type here to simply types upstream
        return templateHTML
    } catch (err) {
        return new Error(err)
    }
}

// SHARED EMAIL LOGIC
// Types

// Constants
// This should reference UUIDS in the statePrograms.json in src/data/
const CHIP_PROGRAMS_UUID = {
    MS: '36c54daf-7611-4a15-8c3b-cdeb3fd7e25a',
    AS: 'e112301b-72c7-4c8f-856a-2cf8c6a1465b',
}

const SubmissionTypeRecord: Record<SubmissionType, string> = {
    CONTRACT_ONLY: 'Contract action only',
    CONTRACT_AND_RATES: 'Contract action and rate certification',
}

// Util Functions
// Checks if at least one program is CHIP
const includesChipPrograms = (programIDs: string[]): boolean => {
    const chipProgramIds = Object.values(CHIP_PROGRAMS_UUID)
    return programIDs.some((id: string) => chipProgramIds.includes(id))
}
// Filter reviewers email list to ensure CHIP programs and state of PR submission do not include OACT and DMCP emails.
const filterChipAndPRSubmissionReviewers = (
    reviewers: string[],
    config: EmailConfiguration
) => {
    const { oactEmails, dmcpEmails } = config

    return reviewers.filter(
        (email) => !dmcpEmails.includes(email) && !oactEmails.includes(email)
    )
}

/* 
    Determine reviewers for a given health plan package and state
    - devReviewTeamEmails added to all emails by default
    - dmcoEmails added to all emails by default
    - dmcpEmails added in both CONTRACT_ONLY and CONTRACT_AND_RATES
    - oactEmails added for CONTRACT_AND_RATES
    
    Return should be wrapped in pruneDuplicate to ensure even if config is added twice, we get unique list of reviewers
*/
const generateCMSReviewerEmails = (
    config: EmailConfiguration,
    pkg: LockedHealthPlanFormDataType | UnlockedHealthPlanFormDataType,
    stateAnalystsEmails: StateAnalystsEmails
): string[] | Error => {
    if (
        pkg.submissionType !== 'CONTRACT_AND_RATES' &&
        pkg.submissionType !== 'CONTRACT_ONLY'
    ) {
        return new Error(
            `generateCMSReviewerEmails does not currently support submission type: ${pkg.submissionType}.`
        )
    }

    const { oactEmails, dmcpEmails, dmcoEmails } = config
    const programIDs = findAllPackageProgramIds(pkg)
    let reviewers: string[] = []

    if (pkg.submissionType === 'CONTRACT_ONLY') {
        // Contract submissions reviewer emails
        reviewers = [
            ...config.devReviewTeamEmails,
            ...stateAnalystsEmails,
            ...dmcoEmails,
            ...dmcpEmails,
        ]
    } else if (pkg.submissionType === 'CONTRACT_AND_RATES') {
        //Contract and rate submissions reviewer emails.
        reviewers = [
            ...config.devReviewTeamEmails,
            ...stateAnalystsEmails,
            ...dmcoEmails,
            ...dmcpEmails,
            ...oactEmails,
        ]
    } else {
        console.error('generateCMSReviewerEmails - unknown package type')
    }

    //Remove OACT and DMCP emails from CHIP or State of PR submissions
    if (includesChipPrograms(programIDs) || pkg.stateCode === 'PR') {
        reviewers = filterChipAndPRSubmissionReviewers(reviewers, config)
    }

    return pruneDuplicateEmails(reviewers)
}

//Finds all package program and rate program ids in a package and combines them into one array removing duplicates.
const findAllPackageProgramIds = (
    pkg: UnlockedHealthPlanFormDataType | LockedHealthPlanFormDataType
): string[] => {
    const programs = [...pkg.programIDs]
    if (pkg.submissionType === 'CONTRACT_AND_RATES' && !!pkg.rateInfos.length) {
        const ratePrograms = pkg.rateInfos.flatMap(
            (rateInfo) => rateInfo.rateProgramIDs
        )
        if (ratePrograms?.length) {
            ratePrograms.forEach(
                (id) => id && !programs.includes(id) && programs.push(id)
            )
        }
    }
    return programs
}

//Find state programs from package programs ids
const findPackagePrograms = (
    pkg: UnlockedHealthPlanFormDataType | LockedHealthPlanFormDataType,
    statePrograms: ProgramType[]
): ProgramType[] | Error => {
    const programIDs = findAllPackageProgramIds(pkg)
    const programs = statePrograms.filter((program) =>
        programIDs.includes(program.id)
    )
    if (!programs || programs.length !== programIDs.length) {
        const errMessage = `Can't find programs ${programIDs} from state ${pkg.stateCode}`
        logError('newPackageCMSEmail', errMessage)
        return new Error(errMessage)
    }

    return programs
}

// Clean out HTML tags from an HTML based template
// this way we still have a text alternative for email client rendering html in plaintext
// plaintext is also referenced for unit testing
const stripHTMLFromTemplate = (template: string) => {
    let formatted = template
    // remove BR tags and replace them with line break
    formatted = formatted.replace(/<br>/gi, '\n')
    formatted = formatted.replace(/<br\s\/>/gi, '\n')
    formatted = formatted.replace(/<br\/>/gi, '\n')

    // remove P and A tags but preserve what's inside of them
    formatted = formatted.replace(/<p.*>/gi, '\n')
    formatted = formatted.replace(/<a.*href="(.*?)".*>(.*?)<\/a>/gi, ' $2 ($1)')
    // everything else
    return formatted.replace(/(<([^>]+)>)/gi, '')
}

export {
    stripHTMLFromTemplate,
    CHIP_PROGRAMS_UUID,
    includesChipPrograms,
    generateCMSReviewerEmails,
    renderTemplate,
    SubmissionTypeRecord,
    findAllPackageProgramIds,
    findPackagePrograms,
    filterChipAndPRSubmissionReviewers,
}
