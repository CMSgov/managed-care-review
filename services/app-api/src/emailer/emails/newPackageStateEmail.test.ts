import {
    testEmailConfig,
    mockContractAmendmentFormData,
    mockContractOnlyFormData,
    mockContractAndRatesFormData,
    mockUser,
    mockMNState,
} from '../../testHelpers/emailerHelpers'
import {
    LockedHealthPlanFormDataType,
    generateRateName,
    packageName,
} from '../../../../app-web/src/common-code/healthPlanFormDataType'
import { newPackageStateEmail } from './index'
import { findPackagePrograms } from '../templateHelpers'

test('to addresses list includes current user', async () => {
    const sub = mockContractOnlyFormData()
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )
    expect(template).toEqual(
        expect.objectContaining({
            toAddresses: expect.arrayContaining([user.email]),
        })
    )
})

test('to addresses list includes all state contacts on submission', async () => {
    const sub: LockedHealthPlanFormDataType = {
        ...mockContractOnlyFormData(),
        stateContacts: [
            {
                name: 'test1',
                titleRole: 'Foo1',
                email: 'test1@example.com',
            },
            {
                name: 'test2',
                titleRole: 'Foo2',
                email: 'test2@example.com',
            },
        ],
    }
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )
    sub.stateContacts.forEach((contact) => {
        expect(template).toEqual(
            expect.objectContaining({
                toAddresses: expect.arrayContaining([contact.email]),
            })
        )
    })
})

test('to addresses list does not include duplicate state contacts on submission', async () => {
    const sub: LockedHealthPlanFormDataType = {
        ...mockContractOnlyFormData(),
        stateContacts: [
            {
                name: 'test1',
                titleRole: 'Foo1',
                email: 'test1@example.com',
            },
            {
                name: 'test1',
                titleRole: 'Foo1',
                email: 'test1@example.com',
            },
        ],
    }
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )

    if (template instanceof Error) {
        console.error(template)
        return
    }

    expect(template.toAddresses).toEqual([
        'test+state+user@example.com',
        'test1@example.com',
    ])
})

test('subject line is correct and clearly states submission is complete', async () => {
    const sub = mockContractOnlyFormData()
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const packagePrograms = findPackagePrograms(sub, defaultStatePrograms)

    if (packagePrograms instanceof Error) {
        throw new Error(packagePrograms.message)
    }

    const name = packageName(sub, packagePrograms)

    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )

    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `${name} was successfully submitted.`
            ),
        })
    )
})

test('includes mcog, rate, and team email addresses', async () => {
    const sub = mockContractOnlyFormData()
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const packagePrograms = findPackagePrograms(sub, defaultStatePrograms)

    if (packagePrograms instanceof Error) {
        throw new Error(packagePrograms.message)
    }

    const name = packageName(sub, packagePrograms)

    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )

    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `please reach out to mcog@example.com`
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `please reach out to rates@example.com`
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`${name} was sent to CMS`),
            bodyText: expect.stringContaining(
                `please reach out to mc-review@example.com`
            ),
        })
    )
})

test('includes link to submission', async () => {
    const sub = mockContractAmendmentFormData()
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                `http://localhost/submissions/${sub.id}`
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyHTML: expect.stringContaining(
                `href="http://localhost/submissions/${sub.id}"`
            ),
        })
    )
})

test('includes information about what is next', async () => {
    const sub = mockContractAmendmentFormData()
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining('What comes next:'),
        })
    )
})

test('includes expected data summary for a contract and rates submission State email', async () => {
    const sub: LockedHealthPlanFormDataType = {
        ...mockContractAndRatesFormData(),
        contractDateStart: new Date('01/01/2021'),
        contractDateEnd: new Date('01/01/2025'),
        rateDateStart: new Date('01/01/2021'),
        rateDateEnd: new Date('01/01/2022'),
    }
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const packagePrograms = findPackagePrograms(sub, defaultStatePrograms)

    if (packagePrograms instanceof Error) {
        throw new Error(packagePrograms.message)
    }

    const rateName = generateRateName(sub, packagePrograms)

    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Submission type: Contract action and rate certification'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Rating period: 01/01/2021 to 01/01/2022'
            ),
        })
    )

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Contract effective dates: 01/01/2021 to 01/01/2025'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(rateName),
        })
    )
})

test('includes expected data summary for a rate amendment submission State email', async () => {
    const sub: LockedHealthPlanFormDataType = {
        ...mockContractAndRatesFormData(),
        rateType: 'AMENDMENT',
        contractDateStart: new Date('01/01/2021'),
        contractDateEnd: new Date('01/01/2025'),
        rateDateStart: new Date('01/01/2021'),
        rateDateEnd: new Date('01/01/2022'),
        rateAmendmentInfo: {
            effectiveDateStart: new Date('06/05/2021'),
            effectiveDateEnd: new Date('12/31/2021'),
        },
    }
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const packagePrograms = findPackagePrograms(sub, defaultStatePrograms)

    if (packagePrograms instanceof Error) {
        throw new Error(packagePrograms.message)
    }

    const rateName = generateRateName(sub, packagePrograms)

    const template = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Submission type: Contract action and rate certification'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Rate amendment effective dates: 06/05/2021 to 12/31/2021'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(rateName),
        })
    )
})

test('renders overall email as expected', async () => {
    const sub: LockedHealthPlanFormDataType = {
        ...mockContractAndRatesFormData(),
        rateType: 'AMENDMENT',
        contractDateStart: new Date('2021-01-01'),
        contractDateEnd: new Date('2021-12-31'),
        rateDateStart: new Date('2021-02-02'),
        rateDateEnd: new Date('2021-11-31'),
        rateDateCertified: new Date('2020-12-01'),
        rateAmendmentInfo: {
            effectiveDateStart: new Date('06/05/2021'),
            effectiveDateEnd: new Date('12/31/2021'),
        },
    }
    const user = mockUser()
    const defaultStatePrograms = mockMNState().programs
    const result = await newPackageStateEmail(
        sub,
        user,
        testEmailConfig,
        defaultStatePrograms
    )
    if (result instanceof Error) {
        console.error(result)
        return
    }

    expect(result.bodyHTML).toMatchSnapshot()
})
