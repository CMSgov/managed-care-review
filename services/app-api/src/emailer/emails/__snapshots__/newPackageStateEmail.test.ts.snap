// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders overall email as expected 1`] = `
"MN-new-submission-snapshot was successfully submitted.
<br /><br />
<b>Submission type</b>: Contract action and rate certification<br />
<b>Contract effective dates</b>: 01/01/2021 to 12/31/2021<br />
<b>Rate name</b>: MN-new-submission-snapshot-RATE-20210605-20211231-AMENDMENT-20201201<br />
<b>Rate amendment effective dates</b>: 06/05/2021 to 12/31/2021<br />
<b>Submission description</b>: A submitted submission<br />
<br />
<a href=\\"http://localhost/submissions/test-abc-125\\">View submission</a>
<br /><br />
If you need to make any changes, please contact CMS.
<br /><br />

<div>What comes next:</div>
<ol>
    <li>
        <strong>Check for completeness:</strong> CMS will review all documentation submitted to ensure all required materials were received.
    </li>
    <li>
        <strong>CMS review:</strong> Your submission will be reviewed by CMS for adherence to federal regulations.
    </li>
    <li>
        <strong>Questions:</strong> You may receive questions via email from CMS as they conduct their review.
    </li>
    <li>
        <strong>Decision:</strong> Once all questions have been addressed, CMS will contact you with their final recommendation.

    </li>
</ol>
<br />
<div>If you need assistance or to make changes to your submission:</div>
<ul>
    <li>
        For assistance with programmatic, contractual, or operational issues, please reach out to mcog@example.com and/or your CMS primary contact.
    </li>
    <li>
        For assistance on policy and actuarial issues, please reach out to rates@example.com.
    </li>
    <li>
        For issues related to MC-Review or all other inquiries, please reach out to mc-review@example.com.
    </li>
</ul>
"
`;
